timelock-server/src/test/java/com/palantir/atlasdb/timelock/lock/ExclusiveLockTests.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.lock;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+
+import org.junit.Test;
+
+public class ExclusiveLockTests {
+
+    private static final UUID REQUEST_1 = UUID.randomUUID();
+    private static final UUID REQUEST_2 = UUID.randomUUID();
+    private static final UUID REQUEST_3 = UUID.randomUUID();
+
+    private final ExclusiveLock lock = new ExclusiveLock();
+
+    @Test
+    public void canLockAndUnlock() {
+        lockSynchronously(REQUEST_1);
+        lock.unlock(REQUEST_1);
+    }
+
+    @Test
+    public void lockIsExclusive() {
+        lockSynchronously(REQUEST_1);
+
+        CompletableFuture<Void> future = lockAsync(REQUEST_2);
+        assertNotComplete(future);
+    }
+
+    @Test
+    public void lockCanBeObtainedAfterBeingUnlocked() {
+        lockSynchronously(REQUEST_1);
+        lock.unlock(REQUEST_1);
+
+        lockSynchronously(REQUEST_2);
+    }
+
+    @Test
+    public void queuedRequestObtainsLockAfterBeingUnlocked() {
+        lockSynchronously(REQUEST_1);
+        CompletableFuture<Void> future = lockAsync(REQUEST_2);
+
+        unlock(REQUEST_1);
+
+        assertCompleteSuccessfully(future);
+    }
+
+    @Test
+    public void multipleQueuedRequestsCanObtainLock() {
+        lockSynchronously(REQUEST_1);
+        CompletableFuture<Void> future2 = lockAsync(REQUEST_2);
+        CompletableFuture<Void> future3 = lockAsync(REQUEST_3);
+
+        unlock(REQUEST_1);
+
+        assertCompleteSuccessfully(future2);
+        assertNotComplete(future3);
+
+        unlock(REQUEST_2);
+
+        assertCompleteSuccessfully(future3);
+    }
+
+    @Test
+    public void unlockByNonHolderThrows() {
+        lockSynchronously(REQUEST_1);
+
+        assertThatThrownBy(() -> unlock(UUID.randomUUID()))
+                .isInstanceOf(IllegalStateException.class);
+        assertThat(lock.getCurrentHolder()).isEqualTo(REQUEST_1);
+    }
+
+    @Test
+    public void unlockByWaiterThrows() {
```

Comment:
Hmm not sure we actually care about that case. It's actually a product bug if unlock is ever called by anyone but the holder (which is why we throw `IllegalStateException` rather than returning false or something)